url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/parsertables/FollowSets.java

```diff
@@ -0,0 +1,211 @@
+				// first rule is made programmatically, add Epsilon to FOLLOW
+				put(rule.getRightSymbol(0), Token.EPSILON);
+			}
+			else	{
+				List nonterms = new ArrayList();	// collect receiving nonterminals
+				
+				for (int i = 0; i < rule.rightSize(); i++)	{
+					String symbol = rule.getRightSymbol(i);
+					boolean isTerminal = Token.isTerminal(symbol);
+					
+					if (nonterms.size() > 0)	{	// if there are receivers (not at first element)
+						// add symbol to FOLLOW sets of all nonterminals in local list
+						List l;
+						if (isTerminal)	{
+							l = new ArrayList();
+							l.add(symbol);	// terminal goes to FOLLOW set
+						}
+						else	{
+							l = (List) firstSets.get(symbol);	// FIRST(symbol) goes to FOLLOW set
+						}
+						//System.err.println("rule "+rule+", adding to "+nonterms+" symbols "+v);
+						addToAllFollowSets(nonterms, l);
+					}
+					
+					if (isTerminal || nullAble.isNullable(symbol) == false)	{	// if it is a terminal or can not be null
+						nonterms.clear();	// empty list of receiving nonterminals
+					}
+
+					if (isTerminal == false)	{	// if it is a nonterminal
+						nonterms.add(symbol);
+						
+						if (i == rule.rightSize() - 1)	{	// at end
+							// add left side of rule to FOLLOW sets of all collected nonterminals
+							List l = new ArrayList();
+							l.add(nonterm);
+							addToAllFollowSets(nonterms, l);
+						}
+					}
+				}	// end for all symbols
+			}	// end if rule position 0
+		}	// end for, loop across rules
+	}
+
+	private void addToAllFollowSets(List nonterms, List symbols)	{
+		for (int i = 0; i < nonterms.size(); i++)	{
+			String nt = (String) nonterms.get(i);
+
+			for (int j = 0; j < symbols.size(); j++)	{
+				String s = (String) symbols.get(j);
+
+				if (Nullable.isNull(s) == false)	// do not add empty word
+					put(nt, s);
+			}
+		}
+	}
+
+
+	private void resolveSets()
+		throws ParserBuildException
+	{
+		// Make plain lists of mixed reference lists: all nonterminals are resolved to their terminal lists
+		for (Enumeration e = keys(); e.hasMoreElements(); )	{
+			String key = (String) e.nextElement();
+			//System.err.println("nonterminal "+key+" = "+get(key));
+			List newList = resolveSetSymbol(key, new Hashtable());
+			replace(key, newList);
+		}
+	}
+	
+	private List resolveSetSymbol(String nonterm, Map done)
+		throws ParserBuildException
+	{
+		if (done.get(nonterm) != null)
+			return null;
+		done.put(nonterm, nonterm);	// avoid endless loops
+
+		List values = (List) get(nonterm);
+		List newList = new ArrayList(values.size() * 2);
+		
+		for (int j = 0; j < values.size(); j++)	{
+			String symbol = (String) values.get(j);
+			
+			if (Token.isTerminal(symbol) == false)	{	// resolve nonterminal
+				try	{
+					List l = resolveSetSymbol(symbol, done);
+					for (int i = 0; l != null && i < l.size(); i++)	{
+						String s = (String) l.get(i);
+						if (newList.indexOf(s) < 0)
+							newList.add(s);
+					}
+				}
+				catch (Exception ex)	{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)